### PR TITLE
scripting: add function for clearing rasterizer caches

### PR DESCRIPTION
**Pull request by valentinvanelslande**
This PR adds a scripting function to clear the rasterizer caches.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4532)

### CODE REVIEW

**Comment by FearlessTobi** [comment]
If the cache is cleared every 10 seconds, what's even the point of it then?

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 1]
```diff
@@ -1022,9 +1025,26 @@ void main() {
...
     ASSERT(d24s8_abgr_viewport_u_id != -1);
+
+    if (Settings::values.enable_cache_clear) {
+        auto& timing = Core::System::GetInstance().CoreTiming();
+        constexpr int ClearCacheMs = 10000;
+        cache_clear_event =
+            timing.RegisterEvent("Rasterizer Cache Clear Event", [&timing, this](u64, s64) {
+                FlushAll();
+                while (!surface_cache.empty())
+                    UnregisterSurface(*surface_cache.begin()->second.begin());
```
Comment (wwylele) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: I see that this part is copied from the destructor. You could wrap this into a `ClearAllCache` function

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 2]
```diff
@@ -1022,9 +1025,26 @@ void main() {
...
+                timing.ScheduleEvent(msToCycles(ClearCacheMs), cache_clear_event);
+            });
+        timing.ScheduleEvent(msToCycles(ClearCacheMs), cache_clear_event);
+    }
 }
 
 RasterizerCacheOpenGL::~RasterizerCacheOpenGL() {
+    if (cache_clear_event != nullptr) {
+        Core::System::GetInstance().CoreTiming().UnscheduleEvent(cache_clear_event, 0);
+        cache_clear_event = nullptr;
```
Comment (wwylele) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: No need to reset this as this doesn't do anything in the destructor.

**Comment by valentinvanelslande** [comment]
@FearlessTobi this fixes the low speed when playing for a long time issue

**Comment by valentinvanelslande** [comment]
@FearlessTobi changed to 15 seconds

**Comment by chris062689** [comment]
I think @FearlessTobi is more asking why that amount of time specifically? 
What tests have you ran to determine the cache fills up on lower-end computers and needs to be cleared in that rate of time? 
What steps have you determined that this actually helps on lower-end hardware?

**Comment by valentinvanelslande** [comment]
@chris062689 i only tested it with Intel® Core™ i3-6006U and Intel® Core™ i5-7200U.

**Comment by Regai** [comment]
@valentinvanelslande 

Can you please elaborate more?
Like which games did you test and how was it apparent that 10-15 seconds is a sensible choice?
Why not every 60 seconds, every 5 minutes or 30 minutes for example?

It would be of course desirable to know the usual fillrate of the cache to know when would be a good time to clear it.

**Comment by valentinvanelslande** [comment]
@Regai changed to 1 minute

**Comment by valentinvanelslande** [comment]
> What steps have you determined that this actually helps on lower-end hardware?

it helps. the game slows down after playing for a long time.

> If the cache is cleared every 10 seconds, what's even the point of it then?

now it's 1 minute.

> 10 seconds still seems like a very short and arbitrary value for this.

now it's 1 minute.

> Can you please elaborate more?

tested acnl welcome amiibo and a lot of other games

**Comment by Jules-A** [comment]
How long is "a long time"? How long does it take usually before starting to slow down? Have you done any benchmarks or is it subjective? What do the slowdowns look like, normal fps drop or just stuttering?

**Comment by valentinvanelslande** [comment]
@Jules-A 

> How long does it take usually before starting to slow down?

idk

> What do the slowdowns look like, normal fps drop or just stuttering?

What do the slowdowns look like, normal fps drop or just stuttering?

**Comment by lioncash** [comment]
> What steps have you determined that this actually helps on lower-end hardware?
>
> it helps. the game slows down when playing for a long time.

This doesn't answer their question.

**Comment by valentinvanelslande** [comment]
@lioncash my mom played without and with this hack. it fixes the issue.

**Comment by B3n30** [comment]
Actually I think Citra is missing a cache clear, and I think we discussed that a few weeks ago in discord too, BUT:

- It should probably clear it when a certain size limit is reached rather then a arbitrary time.
- It should probably clear only the outdated cache and not the most recent used one.
- I don't see why - a proper implemented - cache clear needs a option to disable it. (I also think that hidden ini options should be avoided as much as possible)

I'm not sure this is possible in Citras cache implementation.

Clearing the complete cache could result in bad performance issues in games. 

But I agreed with the other reviewers that this change needs to be verified by numbers/tests with/without cache clearing

**Comment by valentinvanelslande** [comment]
can this be added to canary?

**Comment by B3n30** [comment]
No.

Sorry but I just explained above why I don't think this should be integrated into Citra at it's current state. And you did very little to almost nothing to explain why this code should be part of Citra. 

I tried to explain to you a lot in the past that `it fixes the issue.` isn't always enough to get merged.

**Comment by valentinvanelslande** [comment]
changed to a scripting function.

**Comment by valentinvanelslande** [comment]
it's **not** a hack now, @B3n30 can it be added to canary now?

**Comment by B3n30** [comment]
Sorry, but you made it even worse. Now you added a ton of race conditions. Did you even try that this works?

What are your opinions about:

> - It should probably clear it when a certain size limit is reached rather than a arbitrary time.
> - It should probably clear only the outdated cache and not the most recent used one.

And the PR is still missing some numbers like how long does it take the cache to get the big to become an issue? Is this only observed in one game.

**Comment by valentinvanelslande** [comment]
i didn't. waiting for appveyor. this is **not** a proper implementation, is just a simple workaround for users that hate low speed when playing for a long time. this happens with some games, not one. users can clear the cache every x seconds, when a email is received, a discord message, etc. the old implementation doesn't allow that. it only has a setting in the ini files for clearing the cache every minute.

**Comment by valentinvanelslande** [comment]
i'll open a new pr with all issues fixed.

**Pull request closed without merging**